File: frontend/packages/domain/format/src/control/payload.ts
/**
 * What a control object carries inside its AEAD message.
 *
 * The payload is one CBOR map, encrypted as that map followed by zero padding up
 * to its Padmé bucket (FM-11): a control object is one AEAD message, so its
 * stored length is its payload's length, and an unpadded payload would count out
 * for the provider whatever the payload lists — the Entries an Index Snapshot
 * names, the Containers a Keyring maps. This is the meta section's rule (FM-9)
 * applied to control objects.
 *
 * This module owns exactly one of the map's fields — `master_key_epoch`, which
 * every control object carries whatever its kind
 * (FM-13) — and treats the rest as the kind's own business: the caller hands
 * over the CBOR map of its fields, and gets those fields back on the way out.
 * The bytes they come back as need not be the bytes they went in as, for the
 * reason [`ControlPayload.body`] gives.
 */

import {
  asCborMap,
  asUint,
  decodeCborExact,
  decodeCborFirst,
  encodeCbor,
} from '../internal/cbor.js';
import { isAllZero } from '../internal/bytes.js';
import { fail } from '../errors.js';
import { paddedLength } from '../padme.js';
import { MasterKeyEpoch } from '../model/masterKeyEpoch.js';

/** The one payload field the framing itself defines. */
const MASTER_KEY_EPOCH = 'master_key_epoch';

/** A control-object payload: the epoch that wrote it, and the kind's own map. */
export interface ControlPayload {
  /** The Master Key epoch that encrypted this object. */
  masterKeyEpoch: MasterKeyEpoch;
  /**
   * The kind's own fields, as the CBOR map they were serialized to.
   *
   * On the way out the map is re-encoded rather than sliced out of the payload,
   * so a decoded body carries the fields the writer wrote but not necessarily
   * the bytes it wrote them as — this package's encoder orders map keys
   * canonically and a writer need not have. Bodies from two implementations are
   * compared as decoded maps, never as byte strings.
   */
  body: Uint8Array;
}

/** The CBOR spelling of a map with no entries, for a kind with no fields yet. */
export function emptyPayloadBody(): Uint8Array {
  return encodeCbor(new Map(), 'control_payload_encode_failed');
}

/**
 * Serializes a payload to the plaintext that gets encrypted: the kind's own map
 * with `master_key_epoch` added, then zero padding to its Padmé bucket (FM-11).
 */
export function encodeControlPayload(payload: ControlPayload): Uint8Array {
  const map = readPayloadMap(payload.body);
  if (map.has(MASTER_KEY_EPOCH)) {
    fail('malformed_control_payload', `the body already carries ${MASTER_KEY_EPOCH}`);
  }
  map.set(MASTER_KEY_EPOCH, payload.masterKeyEpoch.value);
  return padToBucket(encodeCbor(map, 'control_payload_encode_failed'));
}

/** Grows a payload map to its Padmé bucket with zero bytes (FM-4, FM-11). */
function padToBucket(map: Uint8Array): Uint8Array {
  const padded = paddedLength(BigInt(map.length));
  // Not the `toLength` helper: its `value_out_of_range` is the generic "this
  // reader cannot address that" of a length read off the wire, and this is a
  // length the padding rule itself asks for, which the Rust side raises as
  // `ControlPayloadTooLong`. Both implementations answer it with the same code.
  if (padded > BigInt(Number.MAX_SAFE_INTEGER)) {
    fail(
      'control_payload_too_long',
      `a control-object payload padded to ${padded} bytes is longer than this runtime addresses`,
    );
  }
  const plaintext = new Uint8Array(Number(padded));
  plaintext.set(map, 0);
  return plaintext;
}

/**
 * Parses a payload plaintext, insisting that it says which epoch encrypted it.
 */
export function decodeControlPayload(plaintext: Uint8Array): ControlPayload {
  const map = readPaddedPayloadMap(plaintext);
  if (!map.has(MASTER_KEY_EPOCH)) {
    fail('missing_master_key_epoch', 'control-object payload carries no master_key_epoch');
  }
  const epoch = asUint(map.get(MASTER_KEY_EPOCH), MASTER_KEY_EPOCH, 'malformed_control_payload');
  map.delete(MASTER_KEY_EPOCH);
  return {
    masterKeyEpoch: MasterKeyEpoch.of(epoch),
    body: encodeCbor(map, 'control_payload_encode_failed'),
  };
}

/**
 * Reads the map a payload plaintext carries, holding what follows it to FM-11's
 * padding rule.
 *
 * CBOR is self-delimiting, so nothing records where the map ends: the map is
 * read first and the padding is whatever is left. That tail has to be exactly
 * the zero bytes that carry the map to its Padmé bucket — a plaintext of any
 * other length was written by something that did not pad as the rule says, and a
 * non-zero byte would make the padding a place to ride bytes past a reader.
 */
function readPaddedPayloadMap(plaintext: Uint8Array): Map<unknown, unknown> {
  const [value, padding] = decodeCborFirst(plaintext, 'malformed_control_payload');
  const expected = paddedLength(BigInt(plaintext.length - padding.length));
  if (BigInt(plaintext.length) !== expected) {
    fail(
      'control_padding_length_mismatch',
      `expected a control-object payload padded to ${expected} bytes, found ${plaintext.length}`,
    );
  }
  if (!isAllZero(padding)) {
    fail('non_zero_control_padding', 'control-object payload padding is not zero-filled');
  }
  return asCborMap(value, 'control_payload_not_a_map', 'a control-object payload');
}

/**
 * Reads one CBOR map, rejecting anything else and anything trailing it.
 *
 * This is for the body the caller hands in, which is the kind's own map alone:
 * the padding is the framing's, and it is added once, around the whole payload.
 */
function readPayloadMap(bytes: Uint8Array): Map<unknown, unknown> {
  const value = decodeCborExact(bytes, 'malformed_control_payload');
  return asCborMap(value, 'control_payload_not_a_map', 'a control-object payload');
}

File: backend/crates/domain/coffret-format/src/control/payload.rs
use ciborium::Value;
use coffret_model::MasterKeyEpoch;

use crate::error::{Error, Result};

/// What a control object carries inside its AEAD message.
///
/// The payload is one CBOR map. This module owns exactly one of its fields —
/// `master_key_epoch`, which every control object carries whatever its kind — and
/// treats the rest as the kind's own business: the caller hands over the CBOR map
/// of its fields, and gets that map back on the way out.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ControlPayload {
    /// The Master Key epoch that encrypted this object.
    pub master_key_epoch: MasterKeyEpoch,
    /// The kind's own fields, as the CBOR map they were serialized to.
    pub body: Vec<u8>,
}

/// The one payload field the framing itself defines.
const MASTER_KEY_EPOCH: &str = "master_key_epoch";

impl ControlPayload {
    /// A payload of `body` — a CBOR map — written under `master_key_epoch`.
    pub fn new(master_key_epoch: MasterKeyEpoch, body: Vec<u8>) -> Self {
        Self {
            master_key_epoch,
            body,
        }
    }

    /// A payload carrying nothing but the epoch, for a kind with no fields yet.
    pub fn empty(master_key_epoch: MasterKeyEpoch) -> Self {
        Self::new(master_key_epoch, empty_map())
    }
}

/// The CBOR spelling of a map with no entries.
fn empty_map() -> Vec<u8> {
    to_bytes(&Value::Map(Vec::new())).expect("an empty map always serializes")
}

/// Serializes a payload: the kind's own map with `master_key_epoch` added.
pub(super) fn encode(payload: &ControlPayload) -> Result<Vec<u8>> {
    let mut entries = read_map(&payload.body)?;
    if entries
        .iter()
        .any(|(key, _)| key.as_text() == Some(MASTER_KEY_EPOCH))
    {
        return Err(Error::MalformedControlPayload {
            detail: format!("the body already carries {MASTER_KEY_EPOCH}"),
        });
    }
    // The framing's own field goes first, so it is readable without walking the
    // kind's fields.
    entries.insert(
        0,
        (
            Value::Text(MASTER_KEY_EPOCH.to_owned()),
            Value::from(payload.master_key_epoch.get()),
        ),
    );
    to_bytes(&Value::Map(entries))
}

/// Parses a payload, insisting that it says which epoch encrypted it.
pub(super) fn decode(bytes: &[u8]) -> Result<ControlPayload> {
    let mut entries = read_map(bytes)?;
    let position = entries
        .iter()
        .position(|(key, _)| key.as_text() == Some(MASTER_KEY_EPOCH))
        .ok_or(Error::MissingMasterKeyEpoch)?;
    let (_, epoch) = entries.remove(position);
    let epoch = epoch
        .as_integer()
        .and_then(|integer| u64::try_from(integer).ok())
        .ok_or_else(|| Error::MalformedControlPayload {
            detail: format!("{MASTER_KEY_EPOCH} is not an unsigned integer"),
        })?;

    Ok(ControlPayload::new(
        MasterKeyEpoch::new(epoch)?,
        to_bytes(&Value::Map(entries))?,
    ))
}

/// Reads one CBOR map, rejecting anything else and anything trailing it.
fn read_map(bytes: &[u8]) -> Result<Vec<(Value, Value)>> {
    let mut remaining = bytes;
    let value: Value =
        ciborium::from_reader(&mut remaining).map_err(|error| Error::MalformedControlPayload {
            detail: error.to_string(),
        })?;
    if !remaining.is_empty() {
        return Err(Error::MalformedControlPayload {
            detail: format!("{} bytes follow the payload map", remaining.len()),
        });
    }
    match value {
        Value::Map(entries) => Ok(entries),
        _ => Err(Error::ControlPayloadNotAMap),
    }
}

fn to_bytes(value: &Value) -> Result<Vec<u8>> {
    let mut bytes = Vec::new();
    ciborium::into_writer(value, &mut bytes).map_err(|error| {
        Error::ControlPayloadEncodeFailed {
            detail: error.to_string(),
        }
    })?;
    Ok(bytes)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn body(fields: &[(&str, i64)]) -> Vec<u8> {
        let entries = fields
            .iter()
            .map(|(key, value)| (Value::Text((*key).to_owned()), Value::from(*value)))
            .collect();
        to_bytes(&Value::Map(entries)).expect("a map of integers serializes")
    }

    fn epoch(value: u64) -> MasterKeyEpoch {
        MasterKeyEpoch::new(value).expect("the epoch is valid")
    }

    // FM-13: every control-object payload carries `master_key_epoch`, and the
    // epoch round-trips unchanged along with the kind's own fields.
    #[test]
    fn the_epoch_and_the_body_round_trip() {
        let payload = ControlPayload::new(epoch(3), body(&[("records", 2)]));
        let decoded =
            decode(&encode(&payload).expect("encoding succeeds")).expect("it round-trips");
        assert_eq!(decoded, payload);
    }

    // FM-13: the epoch is a field of the payload map itself, not a wrapper
    // around it, so a reader of a future schema finds it beside the kind's own
    // fields.
    #[test]
    fn the_epoch_is_a_field_of_the_payload_map() {
        let bytes = encode(&ControlPayload::new(epoch(5), body(&[("records", 2)])))
            .expect("encoding succeeds");
        let value: Value = ciborium::from_reader(bytes.as_slice()).expect("the payload is CBOR");
        let entries = value.as_map().expect("the payload is a map");
        let keys: Vec<&str> = entries
            .iter()
            .map(|(key, _)| key.as_text().expect("keys are text"))
            .collect();
        assert_eq!(keys, ["master_key_epoch", "records"]);
        assert_eq!(
            entries[0].1.as_integer(),
            Some(ciborium::value::Integer::from(5u64))
        );
    }

    // FM-13: a payload that does not say which epoch encrypted it is rejected.
    #[test]
    fn a_payload_without_the_epoch_is_rejected() {
        let result = decode(&body(&[("records", 2)]));
        assert!(
            matches!(result, Err(Error::MissingMasterKeyEpoch)),
            "expected a payload without an epoch to be rejected, got {result:?}"
        );
    }

    #[test]
    fn a_payload_that_is_not_a_map_is_rejected() {
        let bytes = to_bytes(&Value::Text("not a map".to_owned())).expect("text serializes");
        let result = decode(&bytes);
        assert!(
            matches!(result, Err(Error::ControlPayloadNotAMap)),
            "expected a payload that is not a map to be rejected, got {result:?}"
        );
    }

    // FM-13: epoch numbering starts at 1, so a payload claiming epoch 0 is not
    // one this build can trust to name a Master Key.
    #[test]
    fn an_epoch_below_one_is_rejected() {
        let bytes = to_bytes(&Value::Map(vec![(
            Value::Text(MASTER_KEY_EPOCH.to_owned()),
            Value::from(0u64),
        )]))
        .expect("the map serializes");
        let result = decode(&bytes);
        assert!(
            matches!(
                result,
                Err(Error::Model(coffret_model::Error::EpochOutOfRange))
            ),
            "expected epoch 0 to be rejected, got {result:?}"
        );
    }

    #[test]
    fn a_non_integer_epoch_is_rejected() {
        let bytes = to_bytes(&Value::Map(vec![(
            Value::Text(MASTER_KEY_EPOCH.to_owned()),
            Value::Text("first".to_owned()),
        )]))
        .expect("the map serializes");
        assert!(matches!(
            decode(&bytes),
            Err(Error::MalformedControlPayload { .. })
        ));
    }

    // The framing owns `master_key_epoch`; a body that also carries one would
    // leave two answers to which epoch wrote the object.
    #[test]
    fn a_body_that_claims_the_epoch_field_is_rejected() {
        let mut payload = ControlPayload::empty(epoch(1));
        payload.body = to_bytes(&Value::Map(vec![(
            Value::Text(MASTER_KEY_EPOCH.to_owned()),
            Value::from(9u64),
        )]))
        .expect("the map serializes");
        assert!(matches!(
            encode(&payload),
            Err(Error::MalformedControlPayload { .. })
        ));
    }

    #[test]
    fn trailing_bytes_after_the_map_are_rejected() {
        let mut bytes = encode(&ControlPayload::empty(epoch(1))).expect("encoding succeeds");
        bytes.push(0x00);
        assert!(matches!(
            decode(&bytes),
            Err(Error::MalformedControlPayload { .. })
        ));
    }

    #[test]
    fn an_empty_payload_carries_only_the_epoch() {
        let payload = ControlPayload::empty(epoch(2));
        let decoded =
            decode(&encode(&payload).expect("encoding succeeds")).expect("it round-trips");
        assert_eq!(decoded, payload);
        assert_eq!(decoded.body, empty_map());
    }
}
